io/blob.go

```diff
@@ -0,0 +1,311 @@
+	if n == 0 && count > 0 {
+		return nil, io.EOF
+	}
+	if count > 0 && n > count {
+		n = count
+	}
+	list := make([]fs.DirEntry, n)
+	for i := range list {
+		list[i] = d.entries[d.offset+i]
+	}
+	d.offset += n
+	return list, nil
+}
+
+func (d *iofsDir) openOnce() error {
+	if d.opened {
+		return nil
+	}
+	d.opened = true
+
+	// blob expects directories to end in the delimiter, except at the top level.
+	prefix := d.key
+	if prefix != "" {
+		prefix += "/"
+	}
+	listOpts := blob.ListOptions{
+		Prefix:    prefix,
+		Delimiter: "/",
+	}
+	ctx := d.b.ctx
+
+	// Fetch all the directory entries.
+	// Conceivably we could only fetch a few here, and fetch the rest lazily
+	// on demand, but that would add significant complexity.
+	iter := d.b.List(&listOpts)
+	for {
+		item, err := iter.Next(ctx)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		name := filepath.Base(item.Key)
+		if item.IsDir {
+			d.entries = append(d.entries, newDir(d.b, item.Key, name))
+		} else {
+			d.entries = append(d.entries, &iofsFileInfo{item, name})
+		}
+	}
+	// There is no such thing as an empty directory in Bucket, so if
+	// we didn't find anything, it doesn't exist.
+	if len(d.entries) == 0 {
+		return fs.ErrNotExist
+	}
+	return nil
+}
+
+// blobOpenFile describes a single open blob as a File.
+// It implements the iceberg-go/io.File interface.
+// It is based on gocloud.dev/blob.iofsOpenFile which:
+// - Doesn't support the `io.ReaderAt` interface
+// - Is not externally accessible, so copied here
+type blobOpenFile struct {
+	*blob.Reader
+	name string
+}
+
+func (f *blobOpenFile) ReadAt(p []byte, off int64) (int, error) {
+	finalOff, err := f.Reader.Seek(off, io.SeekStart)
+	if err != nil {
+		return -1, err
+	} else if finalOff != off {
+		return -1, io.ErrUnexpectedEOF
+	}
+
+	return f.Read(p)
+}
+
+// Functions to implement the `Stat()` function in the `io/fs.File` interface
+
+func (f *blobOpenFile) Name() string               { return f.name }
+func (f *blobOpenFile) Mode() fs.FileMode          { return fs.ModeIrregular }
+func (f *blobOpenFile) Sys() interface{}           { return f.Reader }
+func (f *blobOpenFile) IsDir() bool                { return false }
+func (f *blobOpenFile) Stat() (fs.FileInfo, error) { return f, nil }
+
+// BlobFileIO represents a file system backed by a bucket in object store. It implements the `iceberg-go/io.FileIO` interface.
+type BlobFileIO struct {
+	*blob.Bucket
+	ctx    context.Context
+	opts   *blob.ReaderOptions
+	prefix string
+}
+
+func (io *BlobFileIO) preprocess(n string) string {
+	_, after, found := strings.Cut(n, "://")
+	if found {
+		n = after
+	}
+
+	out := strings.TrimPrefix(n, io.prefix)
+	if out == "/" {
+		out = "."
+	} else {
+		out = strings.TrimPrefix(out, "/")
+	}
+
+	return out
+}
+
+// Open a Blob from a Bucket using the BlobFileIO. Note this
+// function is copied from blob.Bucket.Open, but extended to
+// return a iceberg-go/io.File instance instead of io/fs.File

```

Comment:
could we just wrap and extend the `blob.Bucket.Open` instead of duplicating it here?

Reply:
Looking at it again it has to be a copy because the CDK [`iofsFileInfo`](https://github.com/google/go-cloud/blob/56954848c3aac17e97ce719cfa55fb09ed4e8ddf/blob/blob_fs.go#L37C6-L37C19)  is unexported and doesn't support the `io.ReaderAt` interface .

Reply:
This could be accomplished if gocloud changed this to a `io.ReadSeeker` which would slightly alter the public interface. Otherwise, I can't see a way to do this without copying:

https://github.com/google/go-cloud/blob/0d91a59252ba29091dad4a127be6b654797ff026/blob/blob_fs.go#L54